File: src/api/reportGenerator.ts
/**
 * CAELab V2.0 - 多尺度综合报告生成 API
 * V2.0-007: 多尺度工作流编排器 - 报告生成模块
 */

import { invoke } from '@tauri-apps/api/core'

// ============ 类型定义 ============

/** 报告章节内容类型 */
export type RgContentType = 'text' | 'table' | 'chart' | 'image'

/** 报告章节 */
export interface RgReportSection {
  /** 章节标题 */
  title: string
  /** 关联的分析尺度，null 表示全局章节 */
  scale: string | null
  /** 内容类型 */
  content_type: RgContentType
  /** 章节数据 */
  data: Record<string, unknown>
  /** 排序序号 */
  order: number
}

/** 报告导出格式 */
export type RgReportFormat = 'pdf' | 'html' | 'markdown'

/** 报告生成配置 */
export interface RgReportConfig {
  /** 关联的执行 ID */
  execution_id: string
  /** 报告标题 */
  title: string
  /** 作者 */
  author: string
  /** 报告章节列表 */
  sections: RgReportSection[]
  /** 是否包含方法论章节 */
  include_methodology: boolean
  /** 是否包含原始数据 */
  include_raw_data: boolean
  /** 是否包含参考文献 */
  include_references: boolean
  /** 导出格式 */
  format: RgReportFormat
}

/** 已生成的报告 */
export interface RgGeneratedReport {
  /** 报告 ID */
  report_id: string
  /** 报告标题 */
  title: string
  /** 导出格式 */
  format: string
  /** 文件存储路径 */
  file_path: string
  /** 页数 */
  page_count: number
  /** 生成时间 (ISO 8601) */
  generated_at: string
  /** 文件大小（字节） */
  file_size_bytes: number
  /** 报告章节列表 */
  sections: RgReportSection[]
}

/** 报告模板 */
export interface RgTemplate {
  /** 模板 ID */
  id: string
  /** 模板名称 */
  name: string
  /** 模板描述 */
  description: string
  /** 预定义章节列表 */
  sections: RgReportSection[]
  /** 是否为默认模板 */
  is_default: boolean
}

// ============ API 函数 ============

/**
 * 生成多尺度综合报告
 * @param config - 报告生成配置
 * @returns 生成的报告信息
 */
export async function generateReport(config: RgReportConfig): Promise<RgGeneratedReport> {
  return invoke<RgGeneratedReport>('generate_report', { config })
}

/**
 * 预览报告（返回 HTML 内容）
 * @param config - 报告生成配置
 * @returns HTML 预览内容
 */
export async function previewReport(config: RgReportConfig): Promise<{ html_content: string }> {
  return invoke<{ html_content: string }>('preview_report', { config })
}

/**
 * 获取可用的报告模板列表
 * @returns 报告模板数组
 */
export async function getReportTemplates(): Promise<RgTemplate[]> {
  return invoke<RgTemplate[]>('get_report_templates')
}

/**
 * 保存自定义报告模板
 * @param template - 报告模板
 * @returns 保存后的模板 ID
 */
export async function saveReportTemplate(template: RgTemplate): Promise<{ template_id: string }> {
  return invoke<{ template_id: string }>('save_report_template', { template })
}

/**
 * 获取报告生成历史
 * @param limit - 返回条数上限，默认不限制
 * @returns 历史报告列表
 */
export async function getReportHistory(limit?: number): Promise<RgGeneratedReport[]> {
  return invoke<RgGeneratedReport[]>('get_report_history', { limit })
}

/**
 * 下载已生成的报告文件
 * @param reportId - 报告 ID
 * @returns 文件存储路径
 */
export async function downloadReport(reportId: string): Promise<{ file_path: string }> {
  return invoke<{ file_path: string }>('download_report', { reportId })
}

File: src-tauri/src/commands/report_generator.rs
//! Report Generator Commands (V2.0 - 多尺度工作流编排器)
//!
//! 提供报告生成与管理能力：
//! - 生成/预览报告
//! - 获取/保存报告模板
//! - 获取报告历史
//! - 下载报告
//!
//! Mock: 3 report history entries, 2 templates.

use serde::{Deserialize, Serialize};
use tauri::command;

// ============================================================================
// Data Structures
// ============================================================================

/// 报告
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Report {
    pub id: String,
    pub title: String,
    pub template_id: String,
    pub template_name: String,
    pub content: serde_json::Value,
    pub format: String,
    pub status: String,
    pub comparison_id: Option<String>,
    pub execution_ids: Vec<String>,
    pub created_at: String,
    pub file_size_bytes: Option<usize>,
}

/// 报告模板
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReportTemplate {
    pub id: String,
    pub name: String,
    pub description: String,
    pub sections: Vec<TemplateSection>,
    pub format: String,
    is_builtin: bool,
    pub created_at: String,
    pub updated_at: String,
}

/// 模板章节
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TemplateSection {
    pub id: String,
    pub title: String,
    pub type_field: String,
    pub required: bool,
    pub order: usize,
}

/// 报告历史条目
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReportHistoryEntry {
    pub id: String,
    pub title: String,
    pub template_name: String,
    pub format: String,
    pub status: String,
    pub created_at: String,
    pub file_size_bytes: usize,
}

/// 报告预览
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReportPreview {
    pub id: String,
    pub title: String,
    pub sections: Vec<PreviewSection>,
    pub total_pages: usize,
    pub generated_at: String,
}

/// 预览章节
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewSection {
    pub title: String,
    pub content_preview: String,
    pub word_count: usize,
    pub has_charts: bool,
    pub has_tables: bool,
}

// ============================================================================
// Mock Data
// ============================================================================

fn mock_templates() -> Vec<ReportTemplate> {
    let now = chrono::Utc::now().to_rfc3339();
    vec![
        ReportTemplate {
            id: "tpl-001".to_string(),
            name: "多尺度分析完整报告".to_string(),
            description: "包含所有尺度的详细分析结果、跨尺度关联分析和工程建议的完整报告模板。".to_string(),
            sections: vec![
                TemplateSection {
                    id: "sec-001".to_string(),
                    title: "概述与研究背景".to_string(),
                    type_field: "text".to_string(),
                    required: true,
                    order: 1,
                },
                TemplateSection {
                    id: "sec-002".to_string(),
                    title: "DFT 计算结果".to_string(),
                    type_field: "scale_result".to_string(),
                    required: true,
                    order: 2,
                },
                TemplateSection {
                    id: "sec-003".to_string(),
                    title: "MD 模拟结果".to_string(),
                    type_field: "scale_result".to_string(),
                    required: true,
                    order: 3,
                },
                TemplateSection {
                    id: "sec-004".to_string(),
                    title: "相场模拟结果".to_string(),
                    type_field: "scale_result".to_string(),
                    required: true,
                    order: 4,
                },
                TemplateSection {
                    id: "sec-005".to_string(),
                    title: "有限元分析结果".to_string(),
                    type_field: "scale_result".to_string(),
                    required: true,
                    order: 5,
                },
                TemplateSection {
                    id: "sec-006".to_string(),
                    title: "跨尺度关联分析".to_string(),
                    type_field: "cross_scale".to_string(),
                    required: true,
                    order: 6,
                },
                TemplateSection {
                    id: "sec-007".to_string(),
                    title: "结论与工程建议".to_string(),
                    type_field: "text".to_string(),
                    required: true,
                    order: 7,
                },
            ],
            format: "pdf".to_string(),
            is_builtin: true,
            created_at: now.clone(),
            updated_at: now.clone(),
        },
        ReportTemplate {
            id: "tpl-002".to_string(),
            name: "快速对比摘要".to_string(),
            description: "精简版报告模板，仅包含关键结果摘要和对比表格，适用于快速汇报。".to_string(),
            sections: vec![
                TemplateSection {
                    id: "sec-101".to_string(),
                    title: "研究概述".to_string(),
                    type_field: "text".to_string(),
                    required: true,
                    order: 1,
                },
                TemplateSection {
                    id: "sec-102".to_string(),
                    title: "关键结果对比表".to_string(),
                    type_field: "comparison_table".to_string(),
                    required: true,
                    order: 2,
                },
                TemplateSection {
                    id: "sec-103".to_string(),
                    title: "主要发现".to_string(),
                    type_field: "key_findings".to_string(),
                    required: true,
                    order: 3,
                },
            ],
            format: "pdf".to_string(),
            is_builtin: true,
            created_at: now.clone(),
            updated_at: now,
        },
    ]
}

fn mock_history() -> Vec<ReportHistoryEntry> {
    vec![
        ReportHistoryEntry {
            id: "rpt-001".to_string(),
            title: "Mg-Al合金蠕变多尺度分析报告".to_string(),
            template_name: "多尺度分析完整报告".to_string(),
            format: "pdf".to_string(),
            status: "completed".to_string(),
            created_at: "2026-04-02T16:00:00Z".to_string(),
            file_size_bytes: 2_457_600,
        },
        ReportHistoryEntry {
            id: "rpt-002".to_string(),
            title: "Ni基高温合金疲劳分析快速摘要".to_string(),
            template_name: "快速对比摘要".to_string(),
            format: "pdf".to_string(),
            status: "completed".to_string(),
            created_at: "2026-04-01T10:30:00Z".to_string(),
            file_size_bytes: 819_200,
        },
        ReportHistoryEntry {
            id: "rpt-003".to_string(),
            title: "复合材料界面力学研究报告".to_string(),
            template_name: "多尺度分析完整报告".to_string(),
            format: "pdf".to_string(),
            status: "completed".to_string(),
            created_at: "2026-03-28T14:20:00Z".to_string(),
            file_size_bytes: 3_145_728,
        },
    ]
}

// ============================================================================
// Commands
// ============================================================================

/// 生成报告
#[command]
pub fn generate_report(
    title: String,
    template_id: String,
    comparison_id: Option<String>,
    execution_ids: Vec<String>,
    format: Option<String>,
) -> Result<Report, String> {
    tracing::info!(
        "Generating report '{}' with template {}",
        title, template_id
    );
    let templates = mock_templates();
    let template = templates
        .iter()
        .find(|t| t.id == template_id)
        .unwrap_or(&templates[0]);
    let report = Report {
        id: uuid::Uuid::new_v4().to_string(),
        title,
        template_id: template_id.clone(),
        template_name: template.name.clone(),
        content: serde_json::json!({
            "sections": template.sections.len(),
            "format": format.as_deref().unwrap_or("pdf"),
            "status": "completed"
        }),
        format: format.unwrap_or_else(|| "pdf".to_string()),
        status: "completed".to_string(),
        comparison_id,
        execution_ids,
        created_at: chrono::Utc::now().to_rfc3339(),
        file_size_bytes: Some(2_457_600),
    };
    tracing::info!("Report generated: {} (id={})", report.title, report.id);
    Ok(report)
}

/// 预览报告
#[command]
pub fn preview_report(
    title: String,
    template_id: String,
    _comparison_id: Option<String>,
) -> Result<ReportPreview, String> {
    tracing::info!(
        "Previewing report '{}' with template {}",
        title, template_id
    );
    let preview = ReportPreview {
        id: uuid::Uuid::new_v4().to_string(),
        title,
        sections: vec![
            PreviewSection {
                title: "概述与研究背景".to_string(),
                content_preview: "本报告针对Mg-Al合金体系开展了从第一性原理到宏观有限元的多尺度分析...".to_string(),
                word_count: 350,
                has_charts: false,
                has_tables: false,
            },
            PreviewSection {
                title: "DFT 计算结果".to_string(),
                content_preview: "采用VASP-PAW方法，4x4x4 k点网格，520 eV截断能...".to_string(),
                word_count: 520,
                has_charts: true,
                has_tables: true,
            },
            PreviewSection {
                title: "跨尺度关联分析".to_string(),
                content_preview: "四个尺度的计算结果通过参数映射建立了完整的关联链...".to_string(),
                word_count: 680,
                has_charts: true,
                has_tables: true,
            },
        ],
        total_pages: 12,
        generated_at: chrono::Utc::now().to_rfc3339(),
    };
    tracing::info!("Preview generated: {} pages", preview.total_pages);
    Ok(preview)
}

/// 获取报告模板列表
#[command]
pub fn get_report_templates() -> Result<Vec<ReportTemplate>, String> {
    tracing::info!("Getting report templates");
    let templates = mock_templates();
    tracing::info!("Returned {} templates", templates.len());
    Ok(templates)
}

/// 保存报告模板
#[command]
pub fn save_report_template(template: ReportTemplate) -> Result<ReportTemplate, String> {
    tracing::info!("Saving report template: {}", template.name);
    let mut saved = template.clone();
    saved.updated_at = chrono::Utc::now().to_rfc3339();
    tracing::info!("Saved template: {} (id={})", saved.name, saved.id);
    Ok(saved)
}

/// 获取报告历史
#[command]
pub fn get_report_history(limit: Option<usize>) -> Result<Vec<ReportHistoryEntry>, String> {
    tracing::info!("Getting report history (limit={:?})", limit);
    let limit = limit.unwrap_or(20);
    let history = mock_history();
    tracing::info!("Returned {} history entries", history.len().min(limit));
    Ok(history.into_iter().take(limit).collect())
}

/// 下载报告
#[command]
pub fn download_report(report_id: String) -> Result<serde_json::Value, String> {
    tracing::info!("Downloading report: {}", report_id);
    Ok(serde_json::json!({
        "report_id": report_id,
        "download_url": format!("/api/v1/reports/{}/download", report_id),
        "file_name": format!("report_{}.pdf", report_id),
        "file_size_bytes": 2_457_600,
        "content_type": "application/pdf",
        "expires_at": chrono::Utc::now().to_rfc3339()
    }))
}
